Update Native File System article (#2543)
GoogleChrome/web.dev: src/site/content/en/blog/native-file-system/index.md

> ---

Author: Fixed.

  // Create a FileSystemWritableFileStream to write to.
  const writable = await fileHandle.createWritable();
  // Write the contents of the file to the stream.
  await writable.write(contents);
  // Close the file and write the contents to disk.
  await writable.close();
}
```

Writing data to disk uses a [`FileSystemWritableFileStream`][fs-writablestream],
essentially a [`WritableStream`][writable-stream]. Create the stream by calling
`createWritable()` on the file handle object. When `createWritable()` is
called, Chrome first checks if the user has granted write permission to the
file. If permission to write hasn't been granted, the browser will prompt
the user for permission. If permission isn't granted, `createWritable()`
will throw a `DOMException`, and the app will not be able to write to the
file. In the text editor, these `DOMException`s are handled in the
[`saveFile()`][text-editor-app-js] method.

The `write()` method takes a string, which is what we want for a text editor.
But it can also take a [BufferSource][buffersource], or a [Blob][blob].
Unlike most `WritableStreams`, changes are **not** written to disk until the
stream is closed by calling `close()`.

> 

Reviewer: is it worth having some example that actually uses the fact that this is a WritableStream as well? The current example doesn't really use any stream specific bits, instead it just uses our convenience wrappers around what is provided by WritableStream. Not sure what such an example would look like (perhaps something that pipes the result of fetch() to a file?)

Other thoughts on this area is that perhaps we should call out that this is effectively a seekable writable stream? I.e. it's a normal writable stream, but you can also seek in it.

Author: Added an example using fetch and piping the response, and a sentence about `seek` and `truncate`.

essentially a [`WritableStream`][writable-stream]. Create the stream by calling
`createWritable()` on the file handle object. When `createWritable()` is
called, Chrome first checks if the user has granted write permission to the
file. If permission to write hasn't been granted, the browser will prompt
the user for permission. If permission isn't granted, `createWritable()`
will throw a `DOMException`, and the app will not be able to write to the
file. In the text editor, these `DOMException`s are handled in the
[`saveFile()`][text-editor-app-js] method.

The `write()` method takes a string, which is what we want for a text editor.
But it can also take a [BufferSource][buffersource], or a [Blob][blob].

For example, you can pipe a stream directly to it:

```js
async function writeURLToFile(fileHandle, url) {
  // Create a FileSystemWritableFileStream to write to.
  const writable = await fileHandle.createWritable();
  // Make an HTTP request for the contents.
  const response = await fetch(url);
  // Stream the response into the file.
  await response.body.pipeTo(writable);
  // pipeTo() closes the destination pipe by default, no need to close it.
}
```

> You can also [`seek()`][spec-seek], or [`truncate`][spec-truncate] within the

Reviewer 2: ```suggestion
You can also [`seek()`][spec-seek], or [`truncate()`][spec-truncate] within the
```

stream to update the file at a specific position, or resize the file.

{% Aside 'caution' %}

> Changes are **not** written to disk until the stream is closed, either by
> calling `close()` or if the stream is closed by the pipe.

Reviewer 2: ```suggestion
Changes are **not** written to disk until the stream is closed, either when
the script calls `close()` or the stream is closed by the pipe.
```

Author: I re-wrote this section before I saw your feedback. I think I've addressed it, but maybe?

{% endAside %}

{% Details %}
{% DetailsSummary 'h5' %}

Chrome 81 and earlier

> Support for writable streams was added in Chrome 82, and the previous

Reviewer 2: ```suggestion
Support for writable streams was added in Chrome 83, and the previous
```

Reviewer 2: Check this throughout.

[BufferSource][buffersource], or a [Blob][blob]. Finally, finish writing by
calling `FileSystemWriter.close()`.

{% endDetails %}

### Storing file handles in IndexedDB

Starting in Chrome 82, you can save a file handle to IndexedDB. This makes
it possible to keep a list of recently opened or edited files. For example
you could offer to re-open the last file when the app is opened, or show a
list of recently edited files within your app. You can also `postMessage`
file handles between the same top-level origin, for example between tabs or
from tabs to workers.

### Open a directory and enumerate its contents

To enumerate all files in a directory, call `chooseFileSystemEntries()`

> with the `type` option set to `'openDirectory'`. The user selects a directory

Reviewer: `'openDirectory'` is now `'open-directory'`

Author: D'oh, thought I caught all of these, thanks!

[BufferSource][buffersource], or a [Blob][blob]. Finally, finish writing by
calling `FileSystemWriter.close()`.

{% endDetails %}

### Storing file handles in IndexedDB

> Starting in Chrome 82, you can save a file handle to IndexedDB. This makes

Reviewer 2: ```suggestion
Starting in Chrome 83, you can save a file handle to IndexedDB. This makes
```

> Starting in Chrome 82, you can save a file handle to IndexedDB. This makes

Reviewer 2: Either write out Indexed database or enclose it in backticks.

Author: Can't we just called it IndexedDB? It's not actually the API interface, but the API name.

From [MDN](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API):
> IndexedDB is a low-level API for client-side storage of significant amounts of structured data, including files/blobs.

it possible to keep a list of recently opened or edited files. For example
you could offer to re-open the last file when the app is opened, or show a

> list of recently edited files within your app. You can also `postMessage`
> file handles between the same top-level origin, for example between tabs or
> from tabs to workers.

Reviewer 2: ```suggestion
list of recently edited files within your app. You can also call `postMessage()`
to send file handles between the same top-level origin, for example between
tabs or from tabs to workers.
```

});
```

## What's new/changed? {: #whats-new }

Chrome 82

* Support for [writable streams](#save-to-disk) was added, and the previous

>   method for writing to disk (`FileSystemWriter`) was deprecated.

Reviewer: not sure if "deprecated" is the right word here, we actually removed the previous method.

Author: Changed to removed.

> * File handles can now be serialized and stored in IndexedDB, or `postMessage`ed

Reviewer 3: ```suggestion
* File handles can now be serialized and stored in IndexedDB, or sent via `postMessage`
```

  to other windows or workers within the same origin. Note that permissions
  are not retained between browser sessions. For example, when a browser tab
  is re-opened, and a file handle is obtained from IndexedDB, the user will

>   need to grant permission to read/write to the file again.

Reviewer: it is probably important to mention somewhere that currently chrome won't automatically prompt for read access if you try to do a operation a handle where read permission expired (for example restored from IndexedDB in a new browser session). If `await handle.queryPermission()` returns "prompt" you need to explicitly call `await handle.requestPermission()` before you might be able to use the handle.

Author: Added some content to cover that around line 300

* Added support for `isSameEntry()`, which returns true if two entries
  represent the same file or directory.
* Updated usage indicators to indicate whether the user has granted the domain
  permission to files, including a new read-only indicator.

> 

Reviewer: we also added `resolve()` to figure out if a file/directory is a descendant of a directory, and if so return the relative path from the directory to the entry. (envisioned use case: editor with a "project" opened (a directory) shown in a side bar. Now user opens a file in a file picker, after which the editor might want to highlight the opened file in the project side bar).

Author: Added an item for `resolve()`

## What's currently supported? {: #whats-supported }

We're still working on some of the implementation for the
Native File System API, and not everything in the [spec][spec]
(or [explainer][explainer]) has been completed.

As of Chrome 82, the following functionality is not available, or
doesn't match the spec:

> * Non-atomic writes (i.e. calls to `FileSystemFileHandle.createWritable()`

Reviewer: I'm not sure if we should even still mention this, or if we do we should definitely rephrase it. There is no mention of "inPlace" anymore in the spec (other than a note mentioning that there isn't currently any support for non-atomic writes).

We still don't really have a story/solution for use cases where we do need non-atomic writes though, so not sure what we should mention here.

Author: Removed this whole section for the time being.

[BufferSource][buffersource], or a [Blob][blob]. Finally, finish writing by
calling `FileSystemWriter.close()`.

{% endDetails %}

### Storing file handles in IndexedDB

Starting in Chrome 83, file handles are serializable, which means that you
can save a file handle to IndexedDB, or `postMessage` them between the same
top-level origin.

Saving file handles to IndexedDB means that you can store state, or remember
which files a user was working on. This makes it possible to keep a list of
recently opened or edited files, offer to re-open the last file when the app
is opened, et cetera. In the text editor, I store a list of the 5 most recent

> files the user has  opened, making it easy to access those files again.

Reviewer 4: ```suggestion
files the user has opened, making it easy to access those files again.
```

Extra space

Since permissions are not persisted between sessions, you should verify
whether the user has granted permission to the file using
`queryPermission()`. If they haven't, use `requestPermission()` to
request permission.

In the text editor, I created a `verifyPermission()` method that checks
if the user has already granted permission, and if required makes the request.

```js/6,10
async function verifyPermission(fileHandle, withWrite) {
  const opts = {};
  if (withWrite) {
    opts.writable = true;
  }
  // Check if we already have permission, if so, return true.
  if (await fileHandle.queryPermission(opts) === 'granted') {
    return true;
  }
  // Request permission, if the user grants permission, return true.
  if (await fileHandle.requestPermission(opts) === 'granted') {
    return true;
  }
  // The user did nt grant permission, return false.
  return false;
}
```

> By requesting write permission with the read request, I was reduced

Reviewer 4: ```suggestion
By requesting write permission with the read request, I reduced
```

});
```

## What's new/changed? {: #whats-new }

> Chrome 82

Reviewer 2: ```suggestion
Chrome 83
```

Author: Good catch, will fix.

* Support for [writable streams](#save-to-disk) was added, and the previous
  method for writing to disk (`FileSystemWriter`) was removed.

> * File handles can now be serialized and stored in IndexedDB, or sent via

Reviewer 2: IndexedDB again.

Author: I believe IndexedDB is the correct term - from MDN https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API, or is there something else?

>   `postMessage` to other windows or workers within the same origin. Note that

Reviewer 2: ```suggestion
  `postMessage()` to other windows or workers within the same origin. Note that
```

>   permissions are not retained between browser sessions. For example, when a

Reviewer 2: It seems like information about permissions should be gathered together under its own heading.

Author: This kind of felt like it flowed into one thought to me. If you separate them out, the permission comment doesn't fit in context.

  browser tab is re-opened, and a file handle is obtained from IndexedDB, the

>   user will need to grant permission to read/write to the file again.

Reviewer 2: ```suggestion
  user will need to grant permission to read and write to the file again.
```

> * Added support for [`isSameEntry()`][spec-issameentry], which returns `true`

Reviewer 2: A best practice is for bulleted list to follow the same grammatical structure.

Author: Suggestions on how to fix this one?

Reviewer 2: I'm not sure why it's not showing against this page, but I've done a PR: https://github.com/GoogleChrome/web.dev/pull/2569/files

Author: AWESOME - that makes more sense now, got it. I fixed all the but permissions one.

  if two entries represent the same file or directory.
* Updated usage indicators to indicate whether the user has granted the domain
  permission to files, including a new read-only indicator.
* Added support for [`resolve()`][spec-resolve], which will return the

>   relative path from one entry to another. Especially helpful for multi-file

Reviewer 2: ```suggestion
  relative path from one entry to another. This is especially helpful for multi-file
```
